crates/circuit/src/dag_node.rs

```diff
@@ -0,0 +1,329 @@
+        };
+
+        Ok((
+            DAGOpNode {
+                instruction: CircuitInstruction {
+                    operation: op,
+                    qubits: qargs.unbind(),
+                    clbits: cargs.unbind(),
+                },
+                sort_key: sort_key.unbind(),
+            },
+            DAGNode { _node_id: -1 },
+        ))
+    }
+
+    fn __getstate__(slf: PyRef<Self>, py: Python) -> PyObject {
+        (
+            (slf.instruction.__getstate__(py), &slf.sort_key),
+            (slf.as_ref()._node_id),
+        )
+            .into_py(py)
+    }
+
+    fn __setstate__(mut slf: PyRefMut<Self>, state: &Bound<PyAny>) -> PyResult<()> {
+        let (((operation, qubits, clbits), sort_key), nid): DAGOpNodeState = state.extract()?;
+        slf.instruction = CircuitInstruction {
+            operation,
+            qubits,
+            clbits,
+        };
+        slf.sort_key = sort_key;
+        slf.as_mut()._node_id = nid;
+        Ok(())
+    }
+
+    pub fn __getnewargs__(&self, py: Python<'_>) -> PyObject {
+        (
+            &self.instruction.operation,
+            &self.instruction.qubits,
+            &self.instruction.clbits,
+            None::<PyObject>,
+        )
+            .into_py(py)
+    }
+
+    #[getter]
+    fn get_op(&self, py: Python) -> PyObject {
+        self.instruction.operation.clone_ref(py)
+    }
+
+    #[setter]
+    fn set_op(&mut self, op: PyObject) {
+        self.instruction.operation = op;
+    }
+
+    #[getter]
+    fn get_qargs(&self, py: Python) -> Py<PyTuple> {
+        self.instruction.qubits.clone_ref(py)
+    }
+
+    #[setter]
+    fn set_qargs(&mut self, qargs: Py<PyTuple>) {
+        self.instruction.qubits = qargs;
+    }
+
+    #[getter]
+    fn get_cargs(&self, py: Python) -> Py<PyTuple> {
+        self.instruction.clbits.clone_ref(py)
+    }
+
+    #[setter]
+    fn set_cargs(&mut self, cargs: Py<PyTuple>) {
+        self.instruction.clbits = cargs;
+    }
+
+    /// Returns the Instruction name corresponding to the op for this node
+    #[getter]
+    fn get_name(&self, py: Python) -> PyResult<PyObject> {
+        Ok(self
+            .instruction
+            .operation
+            .bind(py)
+            .getattr(intern!(py, "name"))?
+            .unbind())
+    }
+
+    /// Sets the Instruction name corresponding to the op for this node
+    #[setter]
+    fn set_name(&self, py: Python, new_name: PyObject) -> PyResult<()> {
+        self.instruction
+            .operation
+            .bind(py)
+            .setattr(intern!(py, "name"), new_name)
+    }
+
+    /// Returns a representation of the DAGOpNode
+    fn __repr__(&self, py: Python) -> PyResult<String> {
+        Ok(format!(
+            "DAGOpNode(op={}, qargs={}, cargs={})",
+            self.instruction.operation.bind(py).str()?,
+            self.instruction.qubits.bind(py).str()?,
+            self.instruction.clbits.bind(py).str()?
+        ))
+    }
```

Comment:
It's possibly marginally better to defer to Python-space formatting here if we can, especially if there are potentially any strings hiding in inner objects.

Reply:
We talked about this briefly offline, but I'd accidentally uses `str` here instead of `repr`.

Resolved in 041f3419b06b6f3a8c1a9cb0cc1db0021d357d7c